File: android/app/src/main/java/com/guitaripod/pixie/data/purchases/CreditPurchaseManager.kt
package com.guitaripod.pixie.data.purchases

import android.app.Activity
import android.util.Log
import com.guitaripod.pixie.data.api.PixieApiService
import com.guitaripod.pixie.data.api.model.*
import com.guitaripod.pixie.data.repository.CreditsRepository
import com.revenuecat.purchases.Package
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton
import java.util.Date

@Singleton
class CreditPurchaseManager @Inject constructor(
    private val revenueCatManager: RevenueCatManager,
    private val apiService: PixieApiService,
    private val creditsRepository: CreditsRepository
) {
    companion object {
        private const val TAG = "CreditPurchaseManager"
        
        private val PACKAGE_TO_PACK_ID = mapOf(
            "starter" to "starter",
            "basic" to "basic",
            "popular" to "popular",
            "business" to "business",
            "enterprise" to "enterprise"
        )
    }
    
    val purchaseState = revenueCatManager.purchaseState
    val offerings = revenueCatManager.offerings
    
    suspend fun purchaseCreditPack(
        activity: Activity,
        packageToPurchase: Package
    ): Result<CreditPurchaseResult> {
        revenueCatManager.purchaseCreditPack(activity, packageToPurchase)
        
        return purchaseState.filterNotNull()
            .filter { it !is PurchaseState.Loading }
            .take(1)
            .map { state ->
                when (state) {
                    is PurchaseState.Success -> {
                        val packId = PACKAGE_TO_PACK_ID[state.packageId] ?: state.packageId
                        validateAndRecordPurchase(
                            packId = packId,
                            purchaseToken = state.purchaseToken,
                            productId = state.productId
                        )
                    }
                    is PurchaseState.Error -> {
                        Result.failure(Exception(state.message))
                    }
                    is PurchaseState.Cancelled -> {
                        Result.failure(PurchaseCancelledException())
                    }
                    else -> {
                        Result.failure(Exception("Unexpected purchase state"))
                    }
                }
            }
            .first()
    }
    
    private suspend fun validateAndRecordPurchase(
        packId: String,
        purchaseToken: String,
        productId: String
    ): Result<CreditPurchaseResult> {
        return try {
            val request = RevenueCatPurchaseValidationRequest(
                packId = packId,
                purchaseToken = purchaseToken,
                productId = productId,
                platform = "android"
            )
            
            val response = apiService.validateRevenueCatPurchase(request)
            
            if (response.isSuccessful) {
                response.body()?.let { validationResponse ->
                    if (validationResponse.success) {
                        
                        Result.success(
                            CreditPurchaseResult(
                                purchaseId = validationResponse.purchaseId,
                                credits = validationResponse.creditsAdded,
                                newBalance = validationResponse.newBalance,
                                amountUsd = ""
                            )
                        )
                    } else {
                        Result.failure(Exception("Purchase validation failed"))
                    }
                } ?: Result.failure(Exception("Empty response from server"))
            } else {
                val errorBody = response.errorBody()?.string()
                Log.e(TAG, "Purchase validation failed: $errorBody")
                Result.failure(Exception("Purchase validation failed: ${response.code()}"))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error recording purchase", e)
            Result.failure(e)
        }
    }
    
    suspend fun restorePurchases(): Result<List<RestoredPurchase>> {
        return try {
            val restoredPurchases = mutableListOf<RestoredPurchase>()
            
            revenueCatManager.restorePurchases { customerInfo ->
                customerInfo?.let { info ->
                    info.entitlements.all.forEach { (_, entitlement) ->
                        if (entitlement.isActive) {
                            val productId = entitlement.productIdentifier
                            restoredPurchases.add(
                                RestoredPurchase(
                                    productId = productId,
                                    purchaseDate = entitlement.latestPurchaseDate.toString(),
                                    expirationDate = entitlement.expirationDate?.toString()
                                )
                            )
                        }
                    }
                }
            }
            
            if (restoredPurchases.isNotEmpty()) {
            }
            
            Result.success(restoredPurchases)
        } catch (e: Exception) {
            Log.e(TAG, "Error restoring purchases", e)
            Result.failure(e)
        }
    }
    
    // Backend credit packs - single source of truth
    private val _backendCreditPacks = MutableStateFlow<List<CreditPack>>(emptyList())
    
    init {
        // Fetch credit packs from backend on initialization
        fetchBackendCreditPacks()
    }
    
    private fun fetchBackendCreditPacks() {
        kotlinx.coroutines.GlobalScope.launch {
            try {
                val response = apiService.getCreditPacks()
                if (response.isSuccessful) {
                    response.body()?.let { packsResponse ->
                        _backendCreditPacks.value = packsResponse.packs
                    }
                } else {
                    Log.e(TAG, "Failed to fetch credit packs: ${response.code()}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error fetching backend credit packs", e)
            }
        }
    }
    
    fun getCreditPacksWithPricing(): Flow<List<CreditPackWithPrice>> {
        return combine(
            _backendCreditPacks,
            offerings
        ) { backendPacks, revenueCatOfferings ->
            if (backendPacks.isEmpty()) {
                return@combine emptyList<CreditPackWithPrice>()
            }
            
            if (revenueCatOfferings == null) {
                // RevenueCat not available yet (app needs to be from Play Store)
                return@combine emptyList<CreditPackWithPrice>()
            }
            
            val defaultOffering = revenueCatOfferings.current ?: revenueCatOfferings.all.values.firstOrNull()
            val packages = defaultOffering?.availablePackages ?: emptyList()
            
            backendPacks.mapNotNull { pack ->
                val rcPackage = packages.firstOrNull { 
                    it.identifier == pack.id || 
                    it.product.id == pack.id ||
                    it.product.id == "com.guitaripod.pixie.${pack.id}"
                }
                
                rcPackage?.let {
                    CreditPackWithPrice(
                        creditPack = CreditPack(
                            id = pack.id,
                            name = pack.name,
                            credits = pack.credits,
                            bonusCredits = pack.bonusCredits,
                            priceUsdCents = pack.priceUsdCents,
                            description = pack.description,
                            popular = pack.id == "popular"
                        ),
                        rcPackage = it,
                        localizedPrice = it.product.price.formatted
                    )
                }
            }.sortedBy { it.creditPack.priceUsdCents }
        }
    }
    
    fun resetPurchaseState() {
        revenueCatManager.resetPurchaseState()
    }
    
    fun refreshOfferings() {
        fetchBackendCreditPacks()
        revenueCatManager.fetchOfferings()
    }
}

data class CreditPurchaseResult(
    val purchaseId: String,
    val credits: Int,
    val newBalance: Int,
    val amountUsd: String
)

data class RestoredPurchase(
    val productId: String,
    val purchaseDate: String,
    val expirationDate: String?
)

data class CreditPackWithPrice(
    val creditPack: CreditPack,
    val rcPackage: Package,
    val localizedPrice: String
)

class PurchaseCancelledException : Exception("Purchase cancelled by user")

data class CreditPurchaseRequest(
    val packId: String,
    val paymentProvider: String,
    val paymentId: String,
    val productId: String
)

data class CreditPurchaseResponse(
    val purchaseId: String,
    val credits: Int,
    val newBalance: Int,
    val amountUsd: String,
    val status: String
)

data class RevenueCatPurchaseValidationRequest(
    val packId: String,
    val purchaseToken: String,
    val productId: String,
    val platform: String
)

data class RevenueCatPurchaseValidationResponse(
    val success: Boolean,
    val purchaseId: String,
    val creditsAdded: Int,
    val newBalance: Int
)

File: iOS/Pixie/Pixie/Services/CreditPurchaseManager.swift
import Foundation
import RevenueCat
import Combine

struct CreditPurchaseResult {
    let purchaseId: String
    let credits: Int
    let newBalance: Int
    let amountUsd: String
}

struct RestoredPurchase {
    let productId: String
    let purchaseDate: String
    let expirationDate: String?
}

struct CreditPackWithPrice {
    let creditPack: CreditPack
    let rcPackage: Package
    let localizedPrice: String
}

class PurchaseCancelledException: LocalizedError {
    var errorDescription: String? {
        return "Purchase cancelled by user"
    }
}

struct RevenueCatPurchaseValidationRequest: Codable {
    let packId: String
    let purchaseToken: String
    let productId: String
    let platform: String
    
    enum CodingKeys: String, CodingKey {
        case packId = "pack_id"
        case purchaseToken = "purchase_token"
        case productId = "product_id"
        case platform
    }
}


@MainActor
class CreditPurchaseManager {
    static let shared = CreditPurchaseManager()
    
    private let revenueCatManager = RevenueCatManager.shared
    private let apiService: APIServiceProtocol
    private let creditsViewModel: CreditsViewModel
    
    @Published private(set) var purchaseState: PurchaseState = .idle
    @Published private(set) var offerings: Offerings?
    @Published private(set) var creditPacks: [CreditPack] = []
    
    private var cancellables = Set<AnyCancellable>()
    
    private let packageToPackId: [String: String] = [
        "starter": "starter",
        "basic": "basic",
        "popular": "popular",
        "business": "business",
        "enterprise": "enterprise"
    ]
    
    private init(apiService: APIServiceProtocol = APIService.shared) {
        self.apiService = apiService
        self.creditsViewModel = CreditsViewModel(apiService: apiService)
        setupBindings()
    }
    
    private func setupBindings() {
        revenueCatManager.observePurchaseState { [weak self] state in
            self?.purchaseState = state
        }
        
        revenueCatManager.$offerings
            .sink { [weak self] offerings in
                self?.offerings = offerings
            }
            .store(in: &cancellables)
        
        Task {
            await fetchCreditPacks()
        }
    }
    
    func fetchCreditPacks() async {
        do {
            let response = try await apiService.getCreditPacks()
            await MainActor.run {
                self.creditPacks = response.packs
            }
        } catch {
            print("Error fetching credit packs: \(error)")
        }
    }
    
    func purchaseCreditPack(package: Package) async -> Result<CreditPurchaseResult, Error> {
        do {
            try await revenueCatManager.purchaseCreditPack(package: package)
            
            switch purchaseState {
            case .success(let purchaseToken, let productId, let packageId):
                let packId = packageToPackId[packageId] ?? packageId
                return await validateAndRecordPurchase(
                    packId: packId,
                    purchaseToken: purchaseToken,
                    productId: productId
                )
                
            case .error(let message):
                return .failure(NSError(domain: "CreditPurchase", code: -1, userInfo: [NSLocalizedDescriptionKey: message]))
                
            case .cancelled:
                return .failure(PurchaseCancelledException())
                
            default:
                return .failure(NSError(domain: "CreditPurchase", code: -1, userInfo: [NSLocalizedDescriptionKey: "Unexpected purchase state"]))
            }
        } catch {
            return .failure(error)
        }
    }
    
    private func validateAndRecordPurchase(
        packId: String,
        purchaseToken: String,
        productId: String
    ) async -> Result<CreditPurchaseResult, Error> {
        do {
            let request = RevenueCatPurchaseValidationRequest(
                packId: packId,
                purchaseToken: purchaseToken,
                productId: productId,
                platform: "ios"
            )
            
            let response: RevenueCatPurchaseValidationResponse = try await apiService.validateRevenueCatPurchase(request)
            
            if response.success {
                await creditsViewModel.loadBalance()
                
                return .success(CreditPurchaseResult(
                    purchaseId: response.purchaseId,
                    credits: response.creditsAdded,
                    newBalance: response.newBalance,
                    amountUsd: ""
                ))
            } else {
                return .failure(NSError(domain: "CreditPurchase", code: -1, userInfo: [NSLocalizedDescriptionKey: "Purchase validation failed"]))
            }
        } catch {
            print("Error recording purchase: \(error)")
            return .failure(error)
        }
    }
    
    func restorePurchases() async -> Result<[RestoredPurchase], Error> {
        do {
            let customerInfo = try await revenueCatManager.restorePurchases()
            var restoredPurchases: [RestoredPurchase] = []
            
            for (_, entitlement) in customerInfo.entitlements.all {
                if entitlement.isActive {
                    let productId = entitlement.productIdentifier
                    restoredPurchases.append(RestoredPurchase(
                        productId: productId,
                        purchaseDate: entitlement.latestPurchaseDate?.description ?? "",
                        expirationDate: entitlement.expirationDate?.description
                    ))
                }
            }
            
            return .success(restoredPurchases)
        } catch {
            print("Error restoring purchases: \(error)")
            return .failure(error)
        }
    }
    
    func getCreditPacksWithPricing() -> AnyPublisher<[CreditPackWithPrice], Never> {
        Publishers.CombineLatest($creditPacks, $offerings)
            .map { (backendPacks, revenueCatOfferings) -> [CreditPackWithPrice] in
                guard !backendPacks.isEmpty else { return [] }
                
                let defaultOffering = revenueCatOfferings?.current ?? revenueCatOfferings?.all.values.first
                
                guard let packages = defaultOffering?.availablePackages else { return [] }
                
                var creditPacksWithPrice: [CreditPackWithPrice] = []
                
                for pack in backendPacks {
                    if let rcPackage = packages.first(where: { $0.identifier == pack.id }) {
                        let creditPack = CreditPack(
                            id: pack.id,
                            name: pack.name,
                            credits: pack.credits,
                            priceUsdCents: pack.priceUsdCents,
                            bonusCredits: pack.bonusCredits,
                            description: pack.description
                        )
                        
                        creditPacksWithPrice.append(CreditPackWithPrice(
                            creditPack: creditPack,
                            rcPackage: rcPackage,
                            localizedPrice: rcPackage.storeProduct.localizedPriceString
                        ))
                    }
                }
                
                return creditPacksWithPrice.sorted { $0.creditPack.priceUsdCents < $1.creditPack.priceUsdCents }
            }
            .eraseToAnyPublisher()
    }
    
    func getCreditsForPackage(_ packId: String) -> (total: Int, base: Int, bonus: Int) {
        if let pack = creditPacks.first(where: { $0.id == packId }) {
            let total = pack.credits + pack.bonusCredits
            return (total, pack.credits, pack.bonusCredits)
        }
        
        return (0, 0, 0)
    }
    
    func resetPurchaseState() {
        revenueCatManager.resetPurchaseState()
    }
}